[1단계 - 행운의 로또미션] 동동(김동희) 미션 제출합니다.

bigsaigon333 opened the pull request:
@JaeYeopHan  님 안녕하세요! 동동(김동희)입니다. 

1단계 - 행운의 로또미션 제출합니다!


## 1.  [:fire: 데모사이트 :fire: ](https://bigsaigon333.github.io/react-lotto/)

## 2. [:airplane: VSCODE  in web :airplane: ](https://github1s.com/bigsaigon333/react-lotto/tree/step1)

## 3. 구현한 기능 목록 - 1단계
- [x] `Class Component`를 사용합니다.
- [x] 로또 구입 금액을 입력하면, 금액에 해당하는 로또를 발급해야 한다.
- [x] 로또 1장의 가격은 1,000원이다.
- [x] 소비자는 **자동 구매**만 할 수 있다.
- [x] 복권 번호는 번호보기 토글 버튼을 클릭하면, 볼 수 있어야 한다.
- [x] 결과 확인하기 버튼을 누르면 당첨 통계, 수익률을 모달로 확인할 수 있다.
- [x] 로또 당첨 금액은 고정되어 있는 것으로 가정한다.
- [x] 다시 시작하기 버튼을 누르면 초기화 되서 다시 구매를 시작할 수 있다.

## 4. 🔎 데모 gif
<img src="https://user-images.githubusercontent.com/61097373/115213587-a0093600-a13c-11eb-802f-f76b53805af1.gif" width="480">

yujo11 committed:
chore: react를 사용하기 위한 webpack 초기 세팅

yujo11 committed:
chore: eslint, prettier 세팅

yujo11 committed:
chore: .gitignore 세팅

yujo11 committed:
chore: postCSS, tailwindCSS를 사용하기 위한 환경설정

yujo11 committed:
feat: 로직을 붙이기 위한 전체 App의 목업 작성

bigsaigon333 committed:
chore: webpack Hot Module Replacement 적용을 위한 환경 설정

yujo11 committed:
chore: react문법 검사를 위한 eslint standard plugin 설정

bigsaigon333 committed:
chore: tailwind JIT mode 적용을 위한 환경 설정

bigsaigon333 committed:
chore: Public class fields 사용을 위한 eslint, babel 설정

bigsaigon333 committed:
feat: 로또 구입 기능 구현

yujo11 committed:
feat: 구매한 로또 티켓의 번호를 확인하는 기능 구현

yujo11 committed:
feat: 로또 당첨 번호와 보너스 번호를 입력하는 기능 구현

yujo11 committed:
feat: 당첨 통계와 수익을 보여주는 기능, 다시 시작하는 기능 구현

yujo11 committed:
fix: 모달 창을 닫았을 때 기존에 입력한 당첨 번호가 사라지던 오류 수정

yujo11 committed:
fix: 티켓이 없을 때 번호 보기 토글이 움직이던 오류 수정

yujo11 committed:
feat: 티켓 구입을 1회로 제한하는 기능 구현

yujo11 committed:
feat: 구입 금액 입력값 검증 후 메세지를 보여주는 기능 구현

yujo11 committed:
feat: 웹 접근성 향상을 위한 a11y 플러그인 추가 및 코드 수정

yujo11 committed:
feat: 웹 접근성 향상을 위한 마우스 클릭 및 키보드 입력 이벤트 추가

yujo11 committed:
feat: 당첨 번호 입력시 검증 메세지를 보여주는 기능 구현

yujo11 committed:
fix: 당첨번호 입력시 검증 메세지를 제대로 보여주지 못 하던 오류 수정

yujo11 committed:
feat: 당첨번호 입력시 포커스 된 번호 입력란의 검증 메세지를 우선적으로 표시하는 기능 구현

yujo11 committed:
feat: space 버튼 입력을 통해 다시 시작하는 기능 구현

yujo11 committed:
fix: 다시 시작할 때 구입 금액이 초기화 되지 않던 오류 수정

yujo11 committed:
refactor: 전체적인 코드 리팩토링

1. 파일명 변경
2. 불필요한 라인제거
3. 불필요한 CSS제거
4. Modal Class Component -> Function Component 변경

yujo11 committed:
refactor: tailwind css 설정 변경

bigsaigon333 committed:
chore: github pages 배포를 위한 webpack 설정 추가

bigsaigon333 committed:
fix: 결과 확인하기 화면의 총 수익률 수식을 수정한다

bigsaigon333 committed:
feat: 번호보기 토글이 포커스되었을 때 파란색으로 색깔을 변경하는 기능 추가

bigsaigon333 committed:
fix: 불필요한 App.css 파일 삭제

bigsaigon333 committed:
feat: 타이틀 폰트 사이즈  및 마진 조정

bigsaigon333 committed:
feat: 버튼이 포커스되었을 때 box-shadow 효과 추가

bigsaigon333 committed:
docs: 1단계 기능 구현 목록 갱신

bigsaigon333 commented:
### 질문1
디자인적인 부분을 조금 더 신경쓰고 싶었으나, 너무 일정이 지체되고 있어서 적당한 선에서 마무리하였습니다... 
CSS는 정말 체계없이 try - fail - retry 방식으로 작성중입니다.. 
CSS는 내부 로직보다 시각적으로 잘 보여지기만 하면 일단 OK라고 생각하고 넘어가는 경향이 큰 것 같습니다.
**뭔가 체계적으로 CSS를 배울 수 있는 방법이 혹시 있다면 추천해주실 수 있을까요?**
  
CSS 각각의 property는 하나하나씩 배워가고 있는데, 중복되는 css를 제거하는 방법,  유지보수가 하기 좋게끔 CSS를 작성하는 방법 등을 잘 모르겠습니다.  BEM, OOCSS 등의 개념은 알고 있긴 합니다만... 단순히 네이밍 컨벤션 정도로만 생각하고 겉핥기식으로 조금 도입하려고 하다가 시간적 여유가 없어서 포기하고 그러기를 반복하는 것 같습니다... 
**BEM, OOCSS 등을 좀더 깊게 파볼까요?**
  
일단 이번에는 postCSS + tailwindcss 조합으로 어느정도 만족도 있게 디자인을 구성하긴 하였는데,  
CSS framekwork 로 인한 해결이 아닌,  **CSS 기저에 깔린 철학을 알고 싶습니다...**
  
너무 떠먹여 달라는 질문인 것 같은데, 참고할만한 키워드나 홈페이지, 책자 정도 공유해주시면 감사하겠습니다! :bow: :bow: :bow:

bigsaigon333 commented:
### 질문 2  (바보같은 질문 하나 더 추가해봅니다...)
HTML, CSS, JS, UI/UX 등 모두 프론트엔드 개발자의 소양으로 당연히 갖추어야 할 것이지만, 
굳이 정량적으로 비중을 나눈다면 각각을 어느 정도 중요도를 가지고 공부하면 좋을까요?
제가 CSS를 더 공부해야겠다고 느낀 것은, 시각적으로 보여지는 것을 원하는 대로 구현해내었을때의 만족감(?), 성취감(?)이 가장 크기 때문입니다 ㅎㅎ... 
딱히 정해진 비중은 없고, 그냥 하고 싶은거 위주로 해도 된다고 말씀해주실걸로 감히 예상해보지만, 한번 더 여쭈어봅니다 :laughing:

jbee37142 commented on .vscode/settings.json:
```diff
@@ -0,0 +1,11 @@
+  "editor.defaultFormatter": "esbenp.prettier-vscode",
+  "files.trimTrailingWhitespace": true,
+  "editor.formatOnSave": true,
+  "editor.codeActionsOnSave": {
+    "source.fixAll.eslint": true
+  },
```
이거 너무 중요한데 아주 짱입니다 👍

jbee37142 commented on src/App.jsx:
```diff
@@ -0,0 +1,124 @@
+
+    const table = {
+      Escape: () => this.closeModal(),
+      ' ': () => this.resetState(),
```
`' '`가 뭐예요...????

bigsaigon333 replied:
`' '` 는 spacebar를 눌렀을 때를 특정하기 위한 `event.key` 였습니다...

UX를 곰곰히 고민해보았을때 모달이 떠있을때 spacebar를 누르면 다시 시작하기 버튼이 눌리게 하는 것이 그닥 UX향상에 도움이 되지 않는 것 같아서, 해당 코드는 삭제하였습니다~ :tada:

bigsaigon333 replied:
https://github.com/woowacourse/react-lotto/pull/24/commits/85536be174e8c8ccac22a9d0f0d03a061759e586

jbee37142 commented on src/components/Modal.jsx:
```diff
@@ -0,0 +1,156 @@
+
+  const profit = getProfit(ticketRanks, props.tickets);
+
+  const handleResetClick = props.reset;
+  const handleCloseClick = props.close;
```
props를 다시 정의한 이유가 있나요?

bigsaigon333 replied:
eslint-plugin-react plugin을 적용하였는데, 이벤트 핸들러의 prefix 를 handle로 하지 않으면 계속 에러가 나서 별 고민없이 위와 같이 코드를 작성하였습니다..

해당 eslint  룰의 페이지를 방문해서 찬찬히 읽어보다 보니 해당 룰의 의미를 깨닫게 되었습니다.
- 컴포넌트에서 자체적인 핸들러를 정의할 때는 handle prefix를 붙인다.
- 부모 컴포넌트로부터 핸들러를 props로 전달받을 때에는 on prefix 로 핸들러를 전달받는다.


이를 적용하여 수정하였습니다! :congratulations: 

https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-handler-names.md

bigsaigon333 replied:
# :hand:
한가지 궁금한 것은, 메소드 이름을 보았을때 메소드가 의미하는 바를 명확히 하고 싶어서  
메소드를 정의할 때는 메소드 이름을 `resetState`로 하고, `constructor`에서 bind할 때
`this.handleResetClick = this.resetState.bind(this);`  와 같이 하였는데, 이렇게 하는 건 일반적이지 않은 걸까요?

bigsaigon333 replied:
관련 커밋입니다: https://github.com/woowacourse/react-lotto/pull/24/commits/8ae15b458cc4cfa737659598d8b2a316b6525679

jbee37142 replied:
~함수 컴포넌트에선 bind 할 일이 없긴 한데,~
bind만 해주는 데 이름이 달라서 읽기가 더 힘들어지진 않을까요?
케바케이긴 하겠지만! resetState 경우, 그 코드가 복잡하지도 않을 것 같아서 handleResetClick에서 정의를 해줘도 괜찮아보이네요 ㅎㅎ

jbee37142 commented on src/components/Modal.jsx:
```diff
@@ -0,0 +1,156 @@
+    .map(getWinningRank)
+    .reduce(
+      getWinningRankCount,
+      Array.from({ length: winningTable.length }, () => 0)
```
`Array.prototype.fill`?

bigsaigon333 replied:
아.... 이런 간결한 방법이 있었군요!! 
jbee님 덕분에 고급 메소드를 또 배워갑니다~!!! :+1:

bigsaigon333 replied:
https://github.com/woowacourse/react-lotto/pull/24/commits/3f1a27af7bbdc5d1687e2bc1fc71664afe285d95

jbee37142 commented on src/components/Modal.jsx:
```diff
@@ -0,0 +1,156 @@
+        <div
+          className="modal-close absolute m-4 w-6 top-2 right-2 cursor-pointer"
+          role="button"
+          tabIndex="-1"
```
이건 왜 -1로 해두신건가요?

bigsaigon333 replied:
키보드 tab으로 이동하여 div.modal-close 에 접근한 경우, 테두리를 표시하는 시각적 효과를 주고 싶었습니다.
이를 구현하는 방식으로 svg에 시각적 효과를 주고 div.modal-close는 포커스가 되지 않도록 하는 잘못된 방법을 사용하였습니다.

말씀해주신대로 button element 로 수정하니 원하는 바를 쉽게 달성할 수 있었습니다! 감사합니다~! :bowing_man:

bigsaigon333 replied:
https://github.com/woowacourse/react-lotto/pull/24/commits/da4733a5f669cd068a75e42faab267cf9cc2c3b6

jbee37142 commented on src/components/Modal.jsx:
```diff
@@ -0,0 +1,156 @@
+  return (
+    <div className="modal w-full flex fixed inset-0 bg-black bg-opacity-50" role="dialog" aria-modal="true">
+      <div className="modal-inner p-10 m-auto relative rounded-xl bg-white">
+        <div
```
button element가 더 어울리지 않을까요?

bigsaigon333 replied:
https://github.com/woowacourse/react-lotto/pull/24#discussion_r616285384

bigsaigon333 replied:
https://github.com/woowacourse/react-lotto/pull/24/commits/da4733a5f669cd068a75e42faab267cf9cc2c3b6

jbee37142 commented on src/components/Modal.jsx:
```diff
@@ -0,0 +1,156 @@
+            <path className="close-x" d="M 10,10 L 30,30 M 30,10 L 10,30" />
+          </svg>
+        </div>
+        <h2 className="text-2xl font-semibold text-center mb-4">
```
Modal 역할과 Result를 뿌려주는 부분이 함께 있는가 각각의 컴포넌트로 나눠주면 어떨까요?

bigsaigon333 replied:
Result를 뿌려주는 부분을  WinningResult.jsx 로 추출하였습니다! 
관심사의 분리(?)를 적용한 것 같아 뿌듯하네요 :smile:

bigsaigon333 replied:
https://github.com/woowacourse/react-lotto/pull/24/commits/962b99b08b695dd4b13c3e40fd323b24553535db

jbee37142 replied:
앗, Modal에서 Result 뿌려주는 부분을 단순히 추출만 하는 것으론 Modal 컴포넌트를 재사용할 수 없어보이는데요, 요건 동동님이 좀 더 고민해보면 좋을 것 같아서 코멘트만 하고 넘어갑니다~

jbee37142 commented on src/components/Modal.jsx:
```diff
@@ -0,0 +1,156 @@
+          <span role="img" aria-label="trophy">
+            🏆
+          </span>
+          {' 당첨 통계 '}
```
이상한 띄어쓰기가 있는데 스타일 이슈라고 생각되네요? 이런 부분은 스타일로 해결을 해야 할 것 같아요;;ㅎㅎㅎ

bigsaigon333 replied:
a11y eslint plugin을 적용하면서 이모지를 span tag로 감쌌습니다. 
이로 인해 텍스트와 이모지 사이의 간격을 조정하기 위해 공백문자를 넣었는데, 
말씀해주신대로 이는 스타일의 문제이므로 css로 해결해야되는 거네요 ㅎㅎ...

아무 생각없이 넘어갈 뻔했는데, 짚어주셔서 감사합니다~ :+1:

bigsaigon333 replied:
https://github.com/woowacourse/react-lotto/pull/24/commits/40bdfb90873d3c8f1083668545f509d2911f9086

jbee37142 commented on src/components/WinningNumberForm.jsx:
```diff
@@ -0,0 +1,234 @@
+    const inputValues = [...this.state.winningNumberInputValues, this.state.bonusNumberInputValue];
+    const nonEmptyInputValues = inputValues.filter((inputValue) => inputValue !== '');
+
+    switch (true) {
+      case !this.hasUniqueInputValues(nonEmptyInputValues):
+        return MESSAGE.WINNING_NUMBER.DUPLICATED_NUMBERS;
+      case !this.isAllNumberInRange(nonEmptyInputValues):
+        return MESSAGE.WINNING_NUMBER.OUT_OF_RANGE;
+      case this.hasEmptyInputValues():
+        return MESSAGE.WINNING_NUMBER.REQUIRED_NEXT_INPUT;
+      default:
+        return '';
```
switch-case statement의 의도에 맞지 않는 syntax라고 생각됩니다!

bigsaigon333 replied:
switch 구문은 각 조건 중 하나만 일치하여야 하는, 즉 각 조건의 병렬성이 충족되어야 하는 것으로 이해하고 있습니다.
위의 validation message  산정시의 조건들은 두 개 이상의 참이 존재할 수 있어 병렬이 아니며, 제가 부여한 우선순위가 존재하므로 switch 구문에는 맞지 않는 것으로 이해하였습니다!

따라서 if문 - early return 으로 수정하였습니다 :laughing:

bigsaigon333 replied:
https://github.com/woowacourse/react-lotto/pull/24/commits/4ee81016a5c286d93fefd5db87457814ca239fd5

jbee37142 commented on src/components/WinningNumberForm.jsx:
```diff
@@ -0,0 +1,234 @@
+    isReset: PropTypes.bool.isRequired,
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.initialState = {
+      winningNumberInputValues: Array.from({ length: LOTTO.LENGTH }, () => ''),
+      bonusNumberInputValue: '',
+      validationMessage: MESSAGE.REQUIRE_WINNING_NUMBER_INPUT,
+    };
+
+    this.state = { ...this.initialState };
+    this.resetState = this.resetState.bind(this);
+
+    this.isFormValid = this.isFormValid.bind(this);
+    this.isValidInputValue = this.isValidInputValue.bind(this);
+
+    this.isAllNumberInRange = this.isAllNumberInRange.bind(this);
+    this.hasUniqueInputValues = this.hasUniqueInputValues.bind(this);
+
+    this.isNumberInRange = this.isNumberInRange.bind(this);
+    this.isUniqueInputValue = this.isUniqueInputValue.bind(this);
+
+    this.setValidationMessage = this.setValidationMessage.bind(this);
+    this.getValidationMessage = this.getValidationMessage.bind(this);
+    this.getInputValidationMessage = this.getInputValidationMessage.bind(this);
+    this.getFormValidationMessage = this.getFormValidationMessage.bind(this);
+
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleWinningNumberInputChange = this.handleWinningNumberInputChange.bind(this);
+    this.handleBonusNumberInputChange = this.handleBonusNumberInputChange.bind(this);
+
+    this.handleInputFocus = this.handleInputFocus.bind(this);
+  }
```
너무 거대하다고 생각되는데, 컴포넌트의 핸들러 또는 메소드로 정의를 해야 하는가 고민해보시고 컴포넌트를 의미에 맞게 나눌 수 없나? 도 고민해보시면 좋을 것 같아요~

bigsaigon333 replied:
validator의 경우 각 input의 값들을 전달받아서 계산할 수 있으므로, 굳이 컴포넌트 내에 위치할 필요가 없다고 판단하여, util함수로 분리하였습니다! (https://github.com/woowacourse/react-lotto/pull/24/commits/28265b3f2ef0ec2efa157067f9b5281d6acf8ae8)

또한 당첨번호 입력란과 보너스 번호 입력란의 input tag가 중복되므로, NumberInput 이라는 컴포넌트로 분리하였습니다!(https://github.com/woowacourse/react-lotto/pull/24/commits/61ee557f714b29b967a6bc1d811eb682157e5d0e) 


Top-Down 방식으로 프로그램을 작성하여서, 컴포넌트 분리가 거의 되지 않았었습니다. 이번에 NumberInput 컴포넌트를 추출하면서,  이런 input 태그는 재사용이 가능하니까 컴포넌트를 분리하면 좋겠구나! 라는걸 느낄 수 있어서 좋았습니다 :smile: (https://github.com/woowacourse/react-lotto/pull/24/commits/2c2dc4ae4368b44c03d087f9d04117a2ff1980bd)

jbee37142 reviewed (changes requested):
수고하셨습니다~

---

> 질문 1.

CSS는 Cascading Style Sheets 입니다 😄 
책으로 공부해본 적은 없고 (최근에 CSS 완벽가이드 란 책이 나온 것 같은데 읽어보진 못했네요)
잘 만들어진 디자인 시스템 오픈소스들은 스타일링을 어떻게 하고 있는지 살펴보면 큰 도움이 되실 거라 생각합니다.

> 질문 2.

전부요!ㅋㅋㅋㅋ 질문이 모호해서 무슨 답변을 드려야 할지 잘 모르겠네요 ㅎㅎ 진로나 성장에 대해 고민되는 부분이 있으면 DM주세요~

bigsaigon333 committed:
refactor: 이모지와 텍스트 사이의 간격을 css로 제어한다

bigsaigon333 committed:
refactor: 이벤트 핸들러의 이름을 컨벤션에 맞게 수정한다

bigsaigon333 committed:
refactor: 배열의 초기값 설정 로직을 간결화한다

- Array.prototype.fill 메소드를 이용한다

bigsaigon333 committed:
fix: 모달이 열려있을때 스페이스바를 누르면 다시 시작하는 기능을 제거한다

bigsaigon333 committed:
refactor: 모달을 닫는 버튼의 html tag를 button으로 수정한다

bigsaigon333 committed:
refactor: 당첨결과를 나타내는 컴포넌트를 모달에서 추출한다

bigsaigon333 committed:
refactor: 당첨번호 입력 validation message syntax를 수정한다

- switch 구문으로 validation message를 산정하던 것을 if문으로 수정한다.
- switch 구문은 각 조건의 병렬성이 충족되어야 하나, 각 조건은 병렬이 아니라 우선순위가 존재하므로
switch 구문이 알맞지 않다.

bigsaigon333 committed:
refactor: className props 작성을 간결히 한다

- classnames 모듈을 사용하였음.

bigsaigon333 committed:
feat: 고유한 id를 생성하는 유틸함수를 작성한다

bigsaigon333 committed:
chore: prettier가 build 디렉토리내의 파일은 무시하도록 설정한다

bigsaigon333 committed:
refactor: 당첨번호 입력란의 각 Input을 컴포넌트로 추출한다

bigsaigon333 committed:
refactor: 당첨번호 입력란의 validator를 유틸함수로 추출한다

bigsaigon333 committed:
refactor: 당첨번호 입력란에 추출한 validator와 컴포넌트를 적용한다

bigsaigon333 commented:
궁금한 점이 있습니다!
event handler가 수행하는 기능을 명확히 이름으로 드러내는 것과, 
event handler에는 handle prefix를 붙이는 컨벤션이 충돌하는 것처럼 생각되었습니다. 

따라서 이를 절충해서 
reset버튼을 클릭하였을때 state를 초기화하는 click event handler를 아래와 같이 선언할 때에는 resetState로 정의하고, 
constructor에서 bind할 때는  `this.handleResetClick = this.resetState.bind(this);` 와 같이 하였는데, 이렇게 하는 건 일반적이지 않은 걸까요?


- 선언할 때:  
```jsx
  resetState() {
    this.setState(...this.initialState);
  }
```
- constructor에서 bind할 때
 ```jsx
  constructor(props) {
    // ...
    this.handleResetClick = this.resetState.bind(this);
  }
```

https://github.com/woowacourse/react-lotto/pull/24#discussion_r617534637

관련 커밋입니다: https://github.com/woowacourse/react-lotto/pull/24/commits/8ae15b458cc4cfa737659598d8b2a316b6525679

bigsaigon333 commented:
> 질문 1.
CSS는 Cascading Style Sheets 입니다 smile
책으로 공부해본 적은 없고 (최근에 CSS 완벽가이드 란 책이 나온 것 같은데 읽어보진 못했네요)
잘 만들어진 디자인 시스템 오픈소스들은 스타일링을 어떻게 하고 있는지 살펴보면 큰 도움이 되실 거라 생각합니다.


Chakra UI 를 많이 추천해주시는 것 같던데, 여러  디자인 시스템 오픈소스들을 연구해보도록 하겠습니다! :bowing_man: 


>질문 2.
전부요!ㅋㅋㅋㅋ 질문이 모호해서 무슨 답변을 드려야 할지 잘 모르겠네요 ㅎㅎ 진로나 성장에 대해 고민되는 부분이 있으면 DM주세요~

ㅎㅎ... 우문현답이네요 ㅎㅎ.. 요즘 조금 더 자신만의 색깔을 찾아가자고 생각해서 여러 고민들을 하고 있는데, 
추후 저의 생각들이 정리가 되고 여쭤볼게 생기면, 잘 정리해서 연락 드릴수 있도록 하겠습니다~! 
바쁘신데도 DM 달라고 흔쾌히 말씀해주셔서 정말 감사드립니다 :+1:  역시 갓제이비 :bee:

jbee37142 reviewed (approved):
수고하셨어요~
코멘트 2개 더 남겼는데 확인만 해보세요~
